### PR TITLE
[3.3.5][WIP] The battle for the undercity

### LINKED ISSUE #17953: [3.3.5] Battle for the Undercity
**Description**: Final quest of epic chain quest in dragonblight

**Current behaviour**: Tell us what happens
Both alliance and horde events are not scripted and alliance is lacking one pre quest as well

**Expected behaviour**: Tell us what should happen instead
It should be scripted.. jesus this template sucks dong!

**Branch(es)**: 3.3.5 / 6.x (Select the branch(es) affected by this issue) 
3.3.5

https://www.youtube.com/watch?v=PxvQKCJagoI - alliance
https://www.youtube.com/watch?v=thvM9fd-pJA - horde

Im half way done with this, just opening an issue so i can point my PR to something when i commit that.
If you have any information about this event as in spawns/auras/sniffs maybe? please post ;p

### PR DESCRIPTION
**Pull request by Riztazz**
**Changes proposed:**
Script the damn thing

**Target branch(es):** 3.3.5/6.x
3.3.5

**Issues addressed:** Closes #17953

**Known issues and TODO list:** 
Handling SQL first, will push code soon when i refactor it
- [x] SQLs
- [ ] Scripts
- [ ] Diff tests (a lot of mobs)

A lot of stuff from theatre of dreams group which i was part of, so kudos to those guys as well

### CODE REVIEW

**Comment by Kittnz** [comment]
We don't use PR's for SQL, only if it also contains c++ script changes.

**Comment by Riztazz** [comment]
It will contain scripts, didnt push those yet

**Comment by Riztazz** [comment]
Messing around with things for now, it will take some time to code this puppy

**Comment by tkrokli** [comment]
I saved a copy of some enums from one of the other devs: http://pastebin.com/VLaXJnqH
(it is incomplete, he said something like "I can't share any more because it is private").

**Comment by DDuarte** [comment]
>  "I can't share any more because it is private"

image

**Comment by Riztazz** [comment]
I already have it but thanks

**Comment by Riztazz** [comment]
Ill fix those warning/errors later and rewrite it properly, its just showcase if something like this can be done, if not then i have to figure something else out, waiting for review ;p

File: src/server/scripts/EasternKingdoms/battle_for_the_undercity.h [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    MAX_BATTLING_AREAS
+};
+
+enum Zones
+{
+    ZONE_TRISFAL_GLADES                 = 85,
+    ZONE_UNDERCITY                      = 1497
+};
+
+#endif // BATTLE_FOR_UNDERCITY_
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/battle_for_the_undercity.h]: missing new line

File: src/server/scripts/EasternKingdoms/location_tirisfal_glades.cpp [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    {
+        return new location_trisfal_glades_WorldMapZoneScript();
+    }
+};
+
+
+void AddSC_location_trisfal_glades()
+{
+    new location_trisfal_glades();
+}
```
Comment (tkrokli) [inline src/server/scripts/EasternKingdoms/location_tirisfal_glades.cpp]: Here too, missing NewLine.  :-)

**Comment by Riztazz** [comment]
https://www.youtube.com/watch?v=w7WF8QE4SHA&feature=youtu.be

its coming together;p

**Comment by ccrs** [comment]
if u have a zonescript, use it to send the worldstates instead of thrall/varian script xD

File: sql/updates/world/3.3.5/derpydoo.sql [hunk 3]
```diff
@@ -0,0 +1,623 @@
+-- Battle for the Undercity - Horde
+
+DELETE FROM `location_scripts` WHERE `zone_entry` = 85;
+INSERT INTO `location_scripts` VALUES
+(85, "location_trisfal_glades");
```
Comment (Aokromes) [inline sql/updates/world/3.3.5/derpydoo.sql]: whut? location scripts?

File: sql/updates/world/3.3.5/derpydoo_location.sql [hunk 4]
```diff
@@ -0,0 +1,6 @@
+DROP TABLE IF EXISTS `location_scripts`;
+CREATE TABLE `location_scripts`
```
Comment (Aokromes) [inline sql/updates/world/3.3.5/derpydoo_location.sql]: move this to the 1st row of the other file :P

**Comment by Shauren** [comment]
You need to rethink your location_scripts implementation, few points here
* Storing it in ObjectMgr will not be threadsafe, what about maps that can have multiple instances but IsDungeon() is false? This is true for battleground and arena maps, you are not checking anywhere that they cannot have location_script
* ObjectMgr is not a place to just dump more containers in there (and location doesn't fit there)

File: src/server/game/Entities/Player/Player.cpp [hunk 5]
```diff
@@ -9324,6 +9324,13 @@ void Player::SendInitWorldStates(uint32 zoneid, uint32 areaid)
                 data << uint32(4882) << uint32(0);              // 10 WORLD_STATE_HOR_WAVE_COUNT
             }
             break;
+        case 85:    // Tirisfal Glades
+        case 1497:  // Undercity
+        {
+            if (LocationScript* location = sObjectMgr->GetLocationScript(85))
+                location->FillInitialWorldStates(data);
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: You are sending these world states to everyone here, not only people who are on the quest.

Comment (Riztazz) [inline src/server/game/Entities/Player/Player.cpp]: WorldStates are limited to players in phases only, so they won't see it, unless you're talking about performance then ill bool it later

**Comment by Riztazz** [comment]
I have no time lately to do anything, if someone wants to finish this one, feel free to do so ;p

**Comment by Kittnz** [comment]
Finish it @Riztazz  💃 🥇  :D

File: src/server/scripts/EasternKingdoms/location_tirisfal_glades.cpp [hunk 6]
```diff
@@ -0,0 +1,241 @@
+/*
+* Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
```
Comment (Aokromes) [inline src/server/scripts/EasternKingdoms/location_tirisfal_glades.cpp]: aling

File: src/server/scripts/EasternKingdoms/battle_for_the_undercity.h [hunk 7]
```diff
@@ -0,0 +1,110 @@
+/*
+* Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
```
Comment (Aokromes) [inline src/server/scripts/EasternKingdoms/battle_for_the_undercity.h]: aling

Comment (Rushor) [inline src/server/scripts/EasternKingdoms/battle_for_the_undercity.h]: you mean aline?

Comment (Aokromes) [inline src/server/scripts/EasternKingdoms/battle_for_the_undercity.h]: No:
http://www.wordreference.com/es/translation.asp?tranword=align

**Comment by Riztazz** [comment]
I dont have time to finish this @Kittnz sorry :p you're more than welcome to take it as base and i can send you rest of the code and you can finish this, for now imma close that one

**Pull request closed without merging**